Sources/NimiqClient/NimiqClient.swift

```diff
@@ -0,0 +1,1082 @@
+    public init(scheme: String = "http", user: String = "", password: String = "", host: String = "127.0.0.1", port: Int = 8648, session: URLSession? = nil){
+        self.url = "\(scheme)://\(host):\(port)"
+        self.auth = "\(user):\(password)".data(using: String.Encoding.utf8)!.base64EncodedString()
+        if session != nil {
+            self.session = session!
+        } else {
+            self.session = URLSession.shared
+        }
+    }
+
+    /// Used in all JSONRPC requests to fetch the data.
+    /// - Parameter method: JSONRPC method.
+    /// - Parameter params: Parameters used by the request.
+    /// - Returns: If succesfull, returns the model reperestation of the result, `nil` otherwise.
+    private func call<T:Decodable>(method: String, params: [Any]) throws -> T? {
+        var responseObject: Root<T>? = nil
+        var clientError: Error? = nil
+
+        // make JSON object to send to the server
+        let callObject:[String:Any] = [
+            "jsonrpc": "2.0",
+            "method": method,
+            "params": params,
+            "id": id
+        ]
+
+        // prepare the request
+        let data = try JSONSerialization.data(withJSONObject: callObject, options: [])
+        var request = URLRequest(url: URL(string: url)!)
+        request.httpMethod = "POST"
+        request.httpBody = data
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.addValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue("Basic \(auth)", forHTTPHeaderField: "Authorization")
+        // TODO: find a better way to fix the error when the server terminates the connection prematurely
+        request.addValue("close", forHTTPHeaderField: "Connection")
+
+        let semaphore = DispatchSemaphore(value: 0)
+
+        // send the request
+        let task = session.dataTask(with: request, completionHandler: { data, response, error in
+            if error == nil {
+                // serialize the data into an object
+                do {
+                    responseObject = try JSONDecoder().decode(Root<T>.self, from: data! )
+
+                } catch {
+                    clientError = Error.internalError(error.localizedDescription)
+                }
+            } else {
+                clientError = Error.connectionError(error!.localizedDescription)
+            }
+
+            // signal that the request was completed
+            semaphore.signal()
+        })
+        task.resume()
+
+        // wait for the response
+        semaphore.wait()
+
+        // throw if there are any errors
+        if clientError != nil {
+            throw clientError!
+        }
+
+        if let error = responseObject?.error {
+            throw Error.remoteError("\(error.message) (Code: \(error.code)")
+        }
+
+        // increase the JSONRPC client request id for the next request
+        self.id = self.id + 1
```

Comment:
thought: `id` should be incremented even when an error occurs, since request might be different.

Reply:
I see. I makes sense, I'll check other RPC implementations to see what approach has been taken and will make the corresponding changes.

Reply:
This has been fixed, it now increases the request ID even if the request fails. This was also changed in the Python and C# implementations.